File: app/fsharp-2/wpf/src/PuyoPuyo.WPF/Domain/PuyoPair.fs
namespace PuyoPuyo.Domain

/// ぷよペア
type PuyoPair =
    { X: int
      Y: int
      Puyo1Color: PuyoColor // 軸ぷよ
      Puyo2Color: PuyoColor // 2つ目のぷよ
      Rotation: int } // 0: 上, 1: 右, 2: 下, 3: 左

module PuyoPair =
    /// ぷよペアを作成
    let create (x: int) (y: int) (color1: PuyoColor) (color2: PuyoColor) (rotation: int) : PuyoPair =
        { X = x
          Y = y
          Puyo1Color = color1
          Puyo2Color = color2
          Rotation = rotation }

    /// ぷよペアの各ぷよの位置を取得
    let getPositions (pair: PuyoPair) : (int * int) * (int * int) =
        let pos1 = (pair.X, pair.Y)

        let pos2 =
            match pair.Rotation with
            | 0 -> (pair.X, pair.Y - 1) // 上
            | 1 -> (pair.X + 1, pair.Y) // 右
            | 2 -> (pair.X, pair.Y + 1) // 下
            | 3 -> (pair.X - 1, pair.Y) // 左
            | _ -> (pair.X, pair.Y - 1) // デフォルトは上

        (pos1, pos2)

    /// ランダムなぷよペアを生成
    let createRandom (x: int) (y: int) (rotation: int) : PuyoPair =
        let random = System.Random()
        let colors = [| Red; Green; Blue; Yellow |]
        let color1 = colors.[random.Next(colors.Length)]
        let color2 = colors.[random.Next(colors.Length)]
        create x y color1 color2 rotation

File: app/csharp-2/maui/PuyoPuyoMAUI.Core/Domain/PuyoPair.cs
namespace PuyoPuyoMAUI.Domain;

/// <summary>
/// ぷよペア
/// </summary>
public sealed record PuyoPair
{
    /// <summary>
    /// X 座標（軸ぷよ）
    /// </summary>
    public int X { get; init; }

    /// <summary>
    /// Y 座標（軸ぷよ）
    /// </summary>
    public int Y { get; init; }

    /// <summary>
    /// 軸ぷよの色
    /// </summary>
    public PuyoColor Puyo1Color { get; init; }

    /// <summary>
    /// 2 つ目のぷよの色
    /// </summary>
    public PuyoColor Puyo2Color { get; init; }

    /// <summary>
    /// 回転状態 (0: 上, 1: 右, 2: 下, 3: 左)
    /// </summary>
    public int Rotation { get; init; }

    public PuyoPair(int x, int y, PuyoColor puyo1Color, PuyoColor puyo2Color, int rotation)
    {
        X = x;
        Y = y;
        Puyo1Color = puyo1Color;
        Puyo2Color = puyo2Color;
        Rotation = rotation;
    }

    /// <summary>
    /// ぷよペアの各ぷよの位置を取得
    /// </summary>
    public ((int X, int Y) Pos1, (int X, int Y) Pos2) GetPositions()
    {
        var pos1 = (X, Y);
        var pos2 = Rotation switch
        {
            0 => (X, Y - 1),      // 上
            1 => (X + 1, Y),      // 右
            2 => (X, Y + 1),      // 下
            3 => (X - 1, Y),      // 左
            _ => (X, Y - 1)       // デフォルトは上
        };
        return (pos1, pos2);
    }

    /// <summary>
    /// ランダムなぷよペアを生成
    /// </summary>
    public static PuyoPair CreateRandom(int x, int y, int rotation)
    {
        var random = Random.Shared;
        var colors = new[] { PuyoColor.Red, PuyoColor.Green, PuyoColor.Blue, PuyoColor.Yellow };
        var color1 = colors[random.Next(colors.Length)];
        var color2 = colors[random.Next(colors.Length)];
        return new PuyoPair(x, y, color1, color2, rotation);
    }
}
